Code/GraphMol/Wrap/rdchem.cpp

```diff
@@ -78,28 +79,100 @@ struct PySysErrWrite : std::ostream, std::streambuf {
   void write(char c) {
     buffer += c;
     if (c == '\n') {
-      PySys_WriteStderr("%s", (prefix + buffer).c_str());
+      PySys_WriteStderr("%s%s", prefix.c_str(), buffer.c_str());
       buffer.clear();
     }
   }
 #endif
 };
 
-void RDLogError(const std::string &msg) {
+// std::ostream wrapper around Python's logging module
+struct PyLogStream : std::ostream, std::streambuf {
+  PyObject *logfn = nullptr;
+#ifdef RDK_THREADSAFE_SSS
+  static thread_local std::string buffer;
+#else
+  std::string buffer = "";
+#endif
+
+  PyLogStream(std::string level): std::ostream(this) {
+    PyObject *module = PyImport_ImportModule("logging");
+    PyObject *logger = nullptr;
+
+    if (module != nullptr) {
+      logger = PyObject_CallMethod(module, "getLogger", "s", "rdkit");
+      Py_DECREF(module);
+    }
+
+    if (logger != nullptr) {
+      logfn = PyObject_GetAttrString(logger, level.c_str());
+      Py_DECREF(logger);
+    }
+
+    if (PyErr_Occurred()) {
+      PyErr_Print();
+    }
+  }
+
+  ~PyLogStream() {
+    if (!_Py_IsFinalizing()) {
+      Py_XDECREF(logfn);
+    }
+  }
+
+  int overflow(int c) override {
+    write(c);
+    return 0;
+  }
+
+  void write(char c) {
+    if (logfn == nullptr) {
+      return;
+    }
+
+    if (c == '\n') {
+#ifdef RDK_THREADSAFE_SSS
+      PyGILStateHolder h;
+#endif
+      PyObject *result = PyObject_CallFunction(logfn, "s", buffer.c_str());
+      Py_XDECREF(result);
+      buffer.clear();
+    }
+    else {
+      buffer += c;
+    }
+  }
+};
+
+#ifdef RDK_THREADSAFE_SSS
+thread_local std::string PyLogStream::buffer;
+#endif
+
+void RDLogDebug(const std::string &msg) {
   NOGIL gil;
-  BOOST_LOG(rdErrorLog) << msg.c_str() << std::endl;
+  BOOST_LOG(rdDebugLog) << msg.c_str() << std::endl;
+}
+
+void RDLogInfo(const std::string &msg) {
+  NOGIL gil;
```

Comment:
I note that we don't have any threading tests which make me a bit nervous.

Perhaps we should dispense with the NOGIL entirely here as these will always be called via python and the expectation is that they will write to a python stream.

Reply:
I'm unsure about this one - my current version is just copying what was already implemented for errors and warnings.  There is a possible case where some Python code is using these functions to log via the C++ logger.  Not sure what the performance impacts would be - whatever the overhead is to release/reclaim the GIL vs. whatever the overhead is in dealing with `std::ostream`s (I'm guessing it's about even, but have zero data to back that up).

I'll definitely add some threading tests, though!

Reply:
I'd prefer to deal with changing the GIL situation (if necessary) in a separate PR.
This one already has useful functionality and diving into the rabbit hole which is cross-language thread safety and logging here seems like a bad idea